addons/crm/models/crm_lead.py

```diff
@@ -353,20 +357,28 @@ def _onchange_mobile_validation(self):
             self.mobile = self.phone_format(self.mobile)
 
     def _prepare_values_from_partner(self, partner):
-        """ Get a dictionary with values coming from customer information to
-        copy on a lead. Email_from and phone fields get the current lead
-        values to avoid being reset if customer has no value for them. """
+        """ Get a dictionary with values coming from partner information to
+        copy on a lead. Non-address fields get the current lead
+        values to avoid being reset if partner has no value for them. """
+
+        # Sync all address fields from partner, or none, to avoid mixing them.
+        if any(partner[f] for f in PARTNER_ADDRESS_FIELDS_TO_SYNC):
+            values = {f: partner[f] for f in PARTNER_ADDRESS_FIELDS_TO_SYNC}
+        else:
+            values = {f: self[f] for f in PARTNER_ADDRESS_FIELDS_TO_SYNC}
```

Comment:
not sure there's a real need to add those to `values`, as they're gonna be no-op?

Reply:
I think it's necessary to avoid cache miss, since they are computed field we need to set a value in any case.